# Microbial cycling of sulfated polysaccharides in peritidal pustular mats in Shark Bay, Australia

In an effort to make data workflows and findings more accessible and reproducible within the field, I have made a strong effort to include all scripts and files used within this study. Raw and assembled metagenomic data can be accessed at [Zenodo](dx.doi.org/10.5281/zenodo.3874996).  These scripts can (and should) also be used (and improved!) and adapted to your own datasets, should they be helpful. 

### Types of analyses included:
- Assigning taxonomy with GTDB-Tk
- Building a phylogenomic tree and overlaying information using anvi'o
- Identifying gene presence/absence (using chosen cuttoff values) with BLAST
- Creating a gene presence/absence heatmap in R
- Creating a visual map using python
- Links to additional bioinformatic and data analysis resources



